### PR TITLE
Brave new develop

### PR DESCRIPTION
**Pull request by sheavalentine-mdsol**
This PR is a complete overhaul of Farscape.

There is no rush to merge, but it would be nice to get a stable branch.

@JordiPolo @chad-medi - Your feedback in particular would be useful.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 
-gem 'debugger',      '~> 1.6.6'
+gem 'byebug', '~> 3.5.1' if RUBY_VERSION > '2'
 gem 'yard',          '~> 0.8.5'
 gem 'rake',          '~> 0.9'
 gem 'awesome_print', '~> 1.1.0'
 gem 'redcarpet'
 
-gem 'representors', git: 'https://www.github.com/mdsol/crichton-representors.git', tag: 'v0.0.1.pre'
+gem 'representors', git: 'https://www.github.com/mdsol-share/representors.git', branch: '0-0-stable'
```
Comment (ghost) [inline Gemfile:10]: I think this should be on a tag of a sha.  Farscape is a critical part of our tools now, and it shouldn't have dependencies are moving targets.  It should be dependent on gems that which versions (tag/sha) that are none to work.

File: Gemfile [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 
 group :development, :test do
   gem 'pry'
+  #TODO replace both crichton and crichton_test_service with references to stable branches when ready.
+  gem 'moya', git: 'https://www.github.com/mdsol/moya.git', branch: 'develop'
+  gem 'crichton', git: 'https://www.github.com/mdsol-share/crichton.git', branch: 'develop'
```
Comment (ghost) [inline Gemfile:16]: These should also be linked to a sha or a tag.  If those don't exist, I don't see the point of releasing a critical tool that is built off gems that are shifting sand or not brave enough to tag themselves as "something that works".

File: README.md [hunk 3]
```diff
@@ -91,6 +83,11 @@ filtered_drds = search_transition.invoke do |builder|
 end
 ```
 
+You may also invoke transitions with automatic attribute and parameter matching
+```ruby
+drds.transitions['search'].invoke(search_term: '1812')
```
Comment (ghost) [inline README.md:88]: Just to understand...are transitions names always strings?  (although, the invoke method supports symbols?)

File: WRITING_PLUGINS.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+This is a symbol representing the name of the plugin
+
+#### type
+The type of plugin, this is useful for manipulating multiple plguins simultaneously
+
+#### middleware
+Middleware objects as per [Faraday plugins](https://github.com/lostisland/faraday#writing-middleware)
+
+#### extensions
+A hash whos keys are the Farscape clases to be extended and values are a list of classes which will extend the default functionality of Farscape clases.
```
Comment (ghost) [inline WRITING_PLUGINS.md:19]: 'whos' => "whose"

File: WRITING_PLUGINS.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+## Farscape Plugin Hash
+
+A Farscape plugin is described with a hash with a set of well defined keys
+
+#### name
+This is a symbol representing the name of the plugin
+
+#### type
+The type of plugin, this is useful for manipulating multiple plguins simultaneously
```
Comment (ghost) [inline WRITING_PLUGINS.md:13]: plguins => plugins.  Missing period.

File: WRITING_PLUGINS.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
+# Writing Plugins
+
+Farscape Plugins can be used to add drop-in functionality to Farscape. 
+
+## Farscape Plugin Hash
+
+A Farscape plugin is described with a hash with a set of well defined keys
```
Comment (ghost) [inline WRITING_PLUGINS.md:7]: missing period.

File: WRITING_PLUGINS.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+#### name
+This is a symbol representing the name of the plugin
+
+#### type
+The type of plugin, this is useful for manipulating multiple plguins simultaneously
+
+#### middleware
+Middleware objects as per [Faraday plugins](https://github.com/lostisland/faraday#writing-middleware)
```
Comment (ghost) [inline WRITING_PLUGINS.md:16]: missing period.

File: WRITING_PLUGINS.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
+# Writing Plugins
+
+Farscape Plugins can be used to add drop-in functionality to Farscape. 
+
+## Farscape Plugin Hash
+
+A Farscape plugin is described with a hash with a set of well defined keys
+
+#### name
+This is a symbol representing the name of the plugin
```
Comment (ghost) [inline WRITING_PLUGINS.md:10]: missing period.

File: WRITING_PLUGINS.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+#### middleware
+Middleware objects as per [Faraday plugins](https://github.com/lostisland/faraday#writing-middleware)
+
+#### extensions
+A hash whos keys are the Farscape clases to be extended and values are a list of classes which will extend the default functionality of Farscape clases.
+Possible keys are: `:Agent`, `:HttpClient`, `:SafeRepresentorAgent`, `:RepresentorAgent`, and `:TransitionAgent`
+
+#### default_state
+When registered, is this :enabled or :disabled by default it is :enabled
```
Comment (ghost) [inline WRITING_PLUGINS.md:23]: missing period.

File: WRITING_PLUGINS.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+Farscape.enabled_plugins #=> []
+Farscape.disabled_plugins #=> [{name: :Peacekeeper, type: :sebacean, middleware: [TestMiddleware::NoGetNoProblem]}]
+Farscape.enable!(name: :Peacekeeper)
+Farscape.enabled_plugins #=> [{name: :Peacekeeper, type: :sebacean, middleware: [TestMiddleware::NoGetNoProblem]}]
+Farscape.disabled_plugins #=> []
+```
+
+## Plugins on instances
+
+Plugins can also be modified on a per instance basis.  Agent, SafeRepresentorAgent, and Representor all support #plugins, #enabled_plugins, and #disabled_plugins on a per instance basis.  These behave the same as their associated Farscape class methods, except they communicate which plugins are enabled or diables for that instance only.
```
Comment (ghost) [inline WRITING_PLUGINS.md:75]: these method should be formatted with  markdown's "`" syntax.

File: lib/farscape/agent.rb [hunk 11]
```diff
@@ -1,60 +1,113 @@
...
+      rescue JSON::ParserError
+        representing = response
+      end
+      raise error.new(representing) if error
+      representing
+    end
+
+    # TODO share this information with serialization factory base
+    def get_accept_header(media_type)
+      media_types = { hale: 'application/vnd.hale+json' }
```
Comment (ghost) [inline lib/farscape/agent.rb:46]: can't we source this media type from a constant somewhere?  Surely, this value is used all over this code base?

File: lib/farscape/client/http_client.rb [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        {
+          idempotent: ['PUT', 'DELETE'],
+          unsafe: ['POST', 'PATCH'], # http://tools.ietf.org/html/rfc5789
+          safe: ['GET', 'HEAD', 'OPTIONS', 'TRACE', 'CONNECT']
+        }
+      end
+
+      def dispatch_error(response)
+        errors = Farscape::Exceptions
+        http_code = {
```
Comment (ghost) [inline lib/farscape/client/http_client.rb:67]: why isn't this just a constant?  I'm pretty sure that we don't need to instantiate this hash every time farscape dispatches an error? 

Oh, better yet, don't declare it as an hash.  I see that the whole point of this code is that you're avoiding writing a simple ruby case-statement.  Just use a case statement.   No need to even declare a hash.

File: lib/farscape/helpers/partially_ordered_list.rb [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      else
+        item
+      end
+    end
+    def ==(other)
+      item == other.item
+    end
+  end
+
+  CircularOrderingError = Class.new(StandardError)
```
Comment (ghost) [inline lib/farscape/helpers/partially_ordered_list.rb:21]: Seriously? There's no need for this magic. 

``` ruby
class CircularOrderingError < StandardError; end
```

File: lib/farscape/helpers/partially_ordered_list.rb [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+  attr_accessor :elements, :ordering
+
+  def initialize(&block)
+    raise ArgumentError, "#{self.class}.new requires a block" unless block_given?
```
Comment (ghost) [inline lib/farscape/helpers/partially_ordered_list.rb:26]: `block_given?` is used when it is implicity included in a method.  Meaning, every ruby method accepts implicitly a block....a block does not have to formally declared in the method definitions list of parameters.

And yet, it is formally declared here.  Therefore: it should be if block.nil? , raise error.

File: lib/farscape/helpers/partially_ordered_list.rb [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    end
+  end
+
+  def each(&block)
+    return to_enum unless block_given?
+    if @cached_ary && @cached_ary.size == elements.size
+      @cached_ary.each{ |elt| yield elt.item }
+    else
+      @cached_ary = []
+      unadded = elements.map{ |elt| elt=elt.dup; elt.preceders = elt.preceders.dup; elt }
```
Comment (ghost) [inline lib/farscape/helpers/partially_ordered_list.rb:60]: yeah....style guides when here....I think this should be broken up across lines.

File: lib/farscape/helpers/partially_ordered_list.rb [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+      while unadded.any?
+        i = unadded.find_index { |candidate| candidate.preceders.none? }
```
Comment (ghost) [inline lib/farscape/helpers/partially_ordered_list.rb:62]: `i`  => A single letter variable name?  I thought our industry moved away from that.

File: lib/farscape/representor.rb [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+    def initialize(requested_media_type, response, agent)
+      @agent = agent
+      @response = response
+      @requested_media_type = requested_media_type
+      @representor = deserialize(requested_media_type, response.body)
+      handle_extensions
+    end
+    
+    %w(using omitting).each do |meth|
```
Comment (ghost) [inline lib/farscape/representor.rb:24]: what's so hard about using `def using` or `def omitting`.  Why the meta-programming?  

Oh, and if this trickery is removed, your yard doc will display those methods.

File: lib/farscape/transition.rb [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+      call_options[:headers] = @agent.get_accept_header(@agent.media_type).merge(options.headers || {})
+      call_options[:params] = options.parameters unless options.parameters.blank?
+      call_options[:body] = options.attributes unless options.attributes.blank?
+
+      response = @agent.client.invoke(call_options)
+
+      @agent.find_exception(response)
+    end
+
+    %w(using omitting).each do |meth|
```
Comment (ghost) [inline lib/farscape/transition.rb:34]: needless meta-programming.

**Comment by ghost** [comment]
@sheavalentine-mdsol - Thanks for soliciting my feedback.  However, this PR is massive, and it probably cost me another day to get through....I haven't even got through the tests, and I'm 100% certain that it is missing some tests.

Additionally, I recognize that some of the bad parts of the code were not authored by you...since I have see the same stuff pushed to other projects.  So, how could I expect you to clean up after them?  I don't.  Your time is better spent writing good code, not cleaning up after others.

Finally, I'm pretty sure that all the merges into this "brave_new_develop" branch where already reviewed and merged by other engineers.  Obviously, they had no problems with obstructionist meta-programming, single-variable names, and missing tests.  

So, it's on them to own this branch and merge it into develop.  

Take my comments in that light, and keep on moving.

File: Gemfile (hunk 1 above)
Comment (HonoreDB) [inline Gemfile:10]: The whole stable branch thing is my [Cunning Plan](https://sites.google.com/a/mdsol.com/knowledgebase/home/departments/engineering/on-demand-portfolio/paas-high-performance-teams/mcc-teams-collaboration-page/medidata-clinial-cloud/platform-services-blog/gemfilebestpractices). Note that in any given project, this resolves to a SHA in the Gemfile.lock.

**Comment by HonoreDB** [comment]
@sheavalentine-mdsol should I merge this as-is? @bpontes-mdsol do you have a PR waiting on this?

**Comment by sheavalentine-mdsol** [comment]
Yeah, give me a moment to list some issues.

**Comment by sheavalentine-mdsol** [comment]
@HonoreDB - Merge.

**Pull request merged**